java/org/tartarus/snowball/SnowballProgram.java

```diff
@@ -286,13 +314,41 @@ protected int find_among_b(Among v[])
 	}
     }
 
+    // mini version of ArrayUtil.oversize from lucene, specialized to chars
+    static int oversize(int minTargetSize) {
+	int extra = minTargetSize >> 3;
+	if (extra < 3) {
+	    extra = 3;
+	}
+	int newSize = minTargetSize + extra;
+	return (newSize + 3) & 0x7ffffffc;
+    }
+
```

Comment:
yeah, this was just to avoid the worst-case (versus allocating new `char[]` one-by-one to grow a string). IIRC, it is pretty rare that stemmers increase the length of a string anyway, so it is probably an unnecessary optimization. I can change it to `Arrays.copyOf(old + 1)` like I did for the TestApp to keep it simple?

Reply:
I guess you mean `Arrays.copyOf(current, newLength)`?

Some of the stemmers can return a longer stem, but it's very unusual - the only examples in the test data when working in UTF-8 seem to be from estonian and lovins (and the lovins stemmer is provided mostly for academic and/or historical interest):

```
estonian	jäi	-> jääma
estonian	jäid	-> jääma
estonian	jäite	-> jääma
estonian	jääb	-> jääma
estonian	jääd	-> jääma
estonian	jään	-> jääma
estonian	keeb	-> keesi
estonian	käia	-> käisi
estonian	käib	-> käisi
estonian	käis	-> käisi
estonian	müüa	-> müüsi
estonian	müüb	-> müüsi
estonian	müüs	-> müüsi
estonian	sõi	-> söö
estonian	viia	-> viima
estonian	viib	-> viima
estonian	viin	-> viima
estonian	võib	-> võisi
estonian	võid	-> võisi
estonian	võin	-> võisi
estonian	võis	-> võisi
lovins	absolv	-> absolut
lovins	dissolv	-> dissolut
lovins	resolv	-> resolut
lovins	unresolv	-> unresolut
```

The situation will be slightly different for wide character Unicode (e.g. `sõi -> söö` above is the same length stem as input then; I suspect the german stemmer changing `ß` to `ss` can cause longer stems there in wide characters), but this gives a general idea - out of over 11 million different input words, probably less than 100 get longer and of those it's likely to be a single replacement that grows them.

Hmm, I didn't consider cases where the working string grows longer than the input and then gets trimmed back to be the same length or shorter, but that also isn't going to be common.

Reply:
> I guess you mean `Arrays.copyOf(current, newLength)`?

Yes, sorry for the confusion. `Arrays.copyOf` is definitely better here.

I think the original `oversize` stuff was just silly paranoia. At least when reusing arrays and passing `(char[] text, int length)`, the caller usually has extra space avoid allocations anyway (or can ensure such a thing if they really need to).